File: samples/vector_matrix_operations/vecop.cc
#include <frovedis.hpp>
#include <frovedis/core/vector_operations.hpp>

using namespace frovedis;

template <class T>
void show(const std::string& msg, const T& val) {
  std::cout << msg << val << std::endl;
}

template <class T>
void show(const std::string& msg, 
          const std::vector<T>& vec,
          const int& limit = 10) {
  std::cout << msg; debug_print_vector(vec, limit);
}

int main(int argc, char* argv[]) {
  use_frovedis use(argc, argv);
  try {
    if (argc < 2) 
      throw std::runtime_error("Use syntax: mpirun -np <nproc> ./vecop <input-file>");

    auto vec = make_dvector_loadline<double>(argv[1]).gather();

    time_spent t(INFO);
    auto vsum = vec + vec; // calls vector_add(vec, vec)
    t.show("vector_add: ");

    auto vsub = vec - vec; // calls vector_subtract(vec, vec)
    t.show("vector_subtract: ");

    auto vmul = vec * vec; // calls vector_multiply(vec, vec)
    t.show("vector_multiply: ");

    auto vdiv = vec / vec; // calls vector_divide(vec, vec)
    t.show("vector_divide: ");

    auto vneg = -vec; // calls vector_negative(vec)
    t.show("vector_negative: ");

    auto vsq = vector_square(vec);
    t.show("vector_square: ");

    auto vreducesum = vector_sum(vec);
    t.show("vector_sum: ");

    auto vmean = vector_mean(vec);
    t.show("vector_mean: ");

    auto vscalsum = vector_scaled_sum(vec, 2.0);
    t.show("vector_scaled_sum: ");

    auto vsqsum = vector_squared_sum(vec);
    t.show("vector_squared_sum: ");

    auto vnorm = vector_norm(vec);
    t.show("vector_norm: ");

    auto vdot = vector_dot(vec,  vec);
    t.show("vector_dot: ");

    auto vssd = vector_ssd(vec, vec * 2.0);
    t.show("vector_ssd: ");

    auto vssmd = vector_ssmd(vec);
    t.show("vector_ssmd: ");

    auto vaxpy = vector_axpy(vec, vec, 2.0);
    t.show("vector_axpy: ");

    auto vsort = vector_sort(vec);
    t.show("vector_sort: ");

    std::vector<size_t> uind, uinv, ucnt;
    auto vunq = vector_unique(vec, uind, uinv, ucnt); // for numpy-like unique
    //auto vunq = vector_unique(vec); // for only unique
    t.show("vector_unique: ");
    /*
    show("unique-sample: ", vunq);
    show("unique-count: ", ucnt);
    show("unique-index: ", uind);
    show("unique-inverse: ", uinv);
    */

    // only for vector<int>
    //auto vbincnt = vector_bincount<size_t>(vec);
    //t.show("vector_bincount: ");

    auto vlog = vector_log(vec);
    t.show("vector_log: ");

    auto vnlog = vector_negative_log(vec); // -vector_log(vec)
    t.show("vector_negative_log: ");

    auto vexp = vector_exp(vec);
    t.show("vector_exp: ");

    auto vlogsumexp = vector_logsumexp(vec);
    t.show("vector_logsumexp: ");

    auto vminid = vector_argmin(vec); 
    t.show("vector_argmin: ");

    auto vmin = vector_amin(vec); 
    t.show("vector_amin: ");

    auto vmaxid = vector_argmax(vec); 
    t.show("vector_argmax: ");

    auto vmax = vector_amax(vec); 
    t.show("vector_amax: ");

    auto vnnz = vector_count_nonzero(vec); 
    t.show("vector_count_nonzero: ");

    auto index = vector_find_nonzero(vec);
    t.show("vector_find_nonzero: ");

    auto vbin = vector_binarize(vec, 4.0);
    t.show("vector_binarize: ");

    auto vclip = vector_clip(vec, 2.0, 5.0);
    t.show("vector_clip: ");

    auto vtake = vector_take<double>(vec, vector_arrange<size_t>(vec.size()));
    t.show("vector_take: ");

    auto vcast = vector_astype<int>(vec); 
    t.show("vector_astype: ");

    auto vzeros = vector_zeros<int>(1000);
    t.show("vector_zeros: ");

    auto vones = vector_ones<int>(1000);
    t.show("vector_ones: ");

    auto vfull = vector_full<int>(1000, 5.0);
    t.show("vector_full: ");

    std::cout << "-------- results -------- \n";   
    show("data: ", vec);
    show("data + data: ", vsum);
    show("data - data: ", vsub);
    show("data * data: ", vmul);
    show("data / data: ", vdiv);
    show("-data: ", vneg);
    show("square(data): ", vsq);
    show("sum(data): ", vreducesum);
    show("mean(data): ", vmean);
    show("scaled-sum(data, 2): ", vscalsum);
    show("squared_sum(data): ", vsqsum);
    show("euclidean-norm(data): ", vnorm);
    show("(dot) data . data: ", vdot);
    show("sum-squared-diff(data, 2*data): ", vssd);
    show("sum-squared-mean-diff(data): ", vssmd);
    show("(axpy) 2 * data + data: ", vaxpy);
    show("sort(data): ", vsort);
    show("unique(data): ", vunq);
    //show("bincount(data): ", vbincnt);
    show("log(data): ", vlog);
    show("-log(data): ", vnlog);
    show("exp(data): ", vexp);
    show("logsumexp(data): ", vlogsumexp);
    show("argmin(data): ", vminid);
    show("amin(data): ", vmin);
    show("argmax(data): ", vmaxid);
    show("amax(data): ", vmax);
    show("count_nonzero(data): ", vnnz);
    show("vector_find nonzero index: ", index); 
    show("binarize(data, 4): ", vbin);
    show("clip(data, 2, 5): ", vclip);
    show("take(data[0 : N): ", vtake);
    show("astype<double>(data): ", vcast);
    show("zeros(1000): ", vzeros);
    show("ones(1000): ", vones);
    show("full(1000, 5): ", vfull);
  } 
  catch(std::exception& e) {
    std::cout <<  "exception caught: " << e.what() << std::endl;
  }
  return 0;
}

File: samples/vector_matrix_operations/python/vecop.py
import sys
import time
import numpy as np
import scipy
from sklearn import preprocessing as pp

def show(msg, X):
    #if sys.version_info[0] < 3:
        print(msg), 
        print(X)
    #else:
    #    print(msg) print(X)

class time_spent(object):
    def __init__(self):
        self.stime = time.time()

    def show(self, msg):
        print("%s %.6f sec" % (msg, time.time() - self.stime))
        self.stime = time.time()
        
if (len(sys.argv) < 2):
  raise AttributeError("Expected at least two arguments!")
vec = np.loadtxt(sys.argv[1], dtype=np.float64)

t = time_spent()
vsum = vec + vec
t.show("np.add: ")

vsub = vec - vec
t.show("np.subtract: ")

vmul = vec * vec
t.show("np.multiply: ")

vdiv = vec / vec
t.show("np.divide: ")

vneg = -vec
t.show("np.negative: ")

vsq = np.square(vec)
t.show("np.square: ")

vreducesum = np.sum(vec)
t.show("np.sum: ")

vmean = np.mean(vec)
t.show("np.mean: ")

vscalsum = np.sum(vec * 2)
t.show("np.scaled_sum: ")

vsqsum = np.sum(np.square(vec))
t.show("np.squared_sum: ")

vnorm = np.linalg.norm(vec)
t.show("np.linalg.norm: ")

vdot = np.dot(vec,  vec)
t.show("np.dot: ")

vssd = np.sum(np.square(vec - (vec * 2)))
t.show("np.ssd: ")

vssmd = np.sum(np.square(vec - np.mean(vec)))
t.show("np.ssmd: ")

vaxpy = 2 * vec + vec
t.show("np.axpy: ")

vsort = np.sort(vec)
t.show("np.sort: ")

vunq, uinv = np.unique(vec, return_inverse=True)
t.show("np.unique: ")

#vbincnt = np.bincount(vec)
#t.show("np.bincount: ")

vlog = np.log(vec)
t.show("np.log: ")

vnlog = -np.log(vec)
t.show("np.negative_log: ")

vexp = np.exp(vec)
t.show("np.exp: ")

vlogsumexp = scipy.special.logsumexp(vec)
t.show("scipy.special.logsumexp: ")

vminid = np.argmin(vec) 
t.show("np.argmin: ")

vmin = np.amin(vec) 
t.show("np.amin: ")

vmaxid = np.argmax(vec) 
t.show("np.argmax: ")

vmax = np.amax(vec) 
t.show("np.amax: ")

vnnz = np.count_nonzero(vec) 
t.show("np.count_nonzero: ")

vbin = pp.binarize(vec.reshape(1,-1), 4) 
t.show("sklearn.preprocessing.binarize: ")

vclip = np.clip(vec, 2, 5) 
t.show("np.clip: ")

vtake = np.take(vec, np.arange(len(vec))) 
t.show("np.take: ")

vcast = vec.astype(dtype=np.float64) 
t.show("np.astype: ")

vzeros = np.zeros(1000, dtype=np.int32)
t.show("np.zeros: ")

vones = np.ones(1000, dtype=np.int32)
t.show("np.ones: ")

vfull = np.full(1000, 5, dtype=np.int32)
t.show("np.full: ")

#'''
print("-------- results --------")
show("data: ", vec)
show("data + data: ", vsum) 
show("data - data: ", vsub)
show("data * data: ", vmul)
show("data / data: ", vdiv)
show("-data: ", vneg)
show("square(data): ", vsq)
show("sum(data): ", vreducesum)
show("mean(data): ", vmean)
show("scaled-sum(data, 2): ", vscalsum)
show("squared_sum(data): ", vsqsum)
show("norm(data): ", vnorm)
show("(dot) data . data: ", vdot)
show("sum-squared-diff(data, 2*data): ", vssd)
show("sum-squared-mean-diff(data): ", vssmd)
show("(axpy) 2 * data + data: ", vaxpy)
show("sort(data): ", vsort)
show("unique(data): ", vunq)
#show("bincount(data): ", vbincnt)
show("log(data): ", vlog)
show("-log(data): ", vnlog)
show("exp(data): ", vexp)
show("logsumexp(data): ", vlogsumexp)
show("argmin(data): ", vminid)
show("argmax(data): ", vmaxid)
show("amin(data): ", vmin)
show("amax(data): ", vmax)
show("count_nonzero(data): ", vnnz)
show("binarize(data, 4): ", vbin)
show("clip(data, 2, 5): ", vclip)
show("take(data[0 : N]): ", vtake)
show("astype<double>(data): ", vcast)
show("zeros(1000): ", vzeros)
show("ones(1000): ", vones)
show("full(1000, 5): ", vfull)
#'''
